### ISSUE TITLE
AsyncQdrantClient

### ISSUE THREAD

**Issue by dgcnz** [qdrant/qdrant-client#157]
Currently it is possible to interact with qdrant asynchronously using the low level http and grpc async clients, but it is a bit of a step down from the nice high-level synchronous QdrantClient. Furthermore, it requires developers to mess with grpc's generated code, which might be confusing to some.

In my work I've essentially cloned parts of the QdrantClient and replaced async/await when necessary, which provides other developers essentially the same interface but increases redundancy in our codebase.

It would be nice to have a uniform interface for async and sync clients, so that issues like #155 and #60 don't happen. Is there a plan for adding this functionality?

**Comment by kenn** [comment]
I think this is important — when I read https://github.com/openai/chatgpt-retrieval-plugin I realized Qdrant doesn't get async support while Pinecone, Weaviate, Milvus and Redis are supported. (the public method interface is async but it's all sync inside, which is even worse as it blocks the event loop)

**Comment by rxy1212** [comment]
As qdrant offered a solid document which told users how to request every single api in [here](https://qdrant.tech/documentation/quick_start/), users can easily make an async request by requesting the http interface directly with aiohttp.

**Comment by prrao87** [comment]
@rxy1212 Do you have any links or code examples where the async request is made via aiohttp? So we would just rely on Qdrant's internal REST API and authentication protocols and not build a custom REST layer on our ends?

**Comment by rxy1212** [comment]
@prrao87 Yes, there are several examples that I can provide for you.
```
from aiohttp import ClientSession

from typing import List, Dict


VECTOR_DB_CREATE_URL = 'http://host:6333/collections/{}'
VECTOR_DB_INSERT_URL = 'http://host:6333/collections/{}/points?wait=true'
VECTOR_DB_QUERY_URL = 'http://host:6333/collections/{}/points/search'
VECTOR_DB_COUNT_URL = 'http://host:6333/collections/{}/points/count'
VECTOR_DB_SCROLL_URL = 'http://host:6333/collections/{}/points/scroll'
VECTOR_DB_GET_COLLECTIONS_URL = 'http://host:6333/collections'


async def qdrant_create(
        session: ClientSession, 
        collection_name: str, 
        size: int, 
        distance: str='Cosine'
    ) -> None:
    """create new collection

    Args:
        session (ClientSession): the aiohttp client
        collection_name (str): name of the collection
        size (int): vecoter dimension
        distance (str, optional):   defaults to 'Cosine'.
    """
    json = {"name": collection_name, "vectors": {"size": size, "distance": distance}}
    async with session.put(VECTOR_DB_CREATE_URL.format(collection_name),  json=json) as response:
        data = await response.json(content_type='application/json')
        assert data['status'] == 'ok'


async def qdrant_insert(
        session: ClientSession, 
        collection_name: str, 
        uuid: str, 
        vector: List[float], 
        **payload
    ) -> None:
    """insert points

    Args:
        session (ClientSession):  the aiohttp client
        collection_name (str): name of the collection
        uuid (str): uinque id of the vector
        vector (List[float]): the vector will be inserted
    """
    json = {"points": [{"id": uuid, "vector": vector, "payload": payload}]}
    async with session.put(VECTOR_DB_INSERT_URL.format(collection_name),  json=json) as response:
        data = await response.json(content_type='application/json')
        assert data['status'] == 'ok'


async def qdrant_query(
        session: ClientSession, 
        collection_name: str, 
        query_vector: List[float], 
        limit: int, 
        **payload
    ) -> List[Dict]:
    """search the similar vectors

    Args:
        session (ClientSession): the aiohttp client
        collection_name (str): name of the collection
        query_vector (List[float]): query vector
        limit (int): number of the retrieved points

    Returns:
        List[Dict]: result set
    """
    if payload.get('agent', 0) and payload.get('player', 0):
        json = {
            "vector": query_vector,
            "limit": limit,
            "with_payload": True,
            "filter": {
                "must": [
                    {
                        "key": "agent",
                        "match": {
                            "value": payload['agent']
                        }
                    },
                    {
                        "key": "player",
                        "match": {
                            "value": payload['player']
                        }
                    }
                ]
            }
        }
    else:
        json = {
            "vector": query_vector,
            "limit": limit,
            "with_payload": True
        }

    async with session.post(VECTOR_DB_QUERY_URL.format(collection_name), json=json) as response:
        data = await response.json(content_type='application/json')
        return data['result']


async def qdrant_scroll(
        session: ClientSession,
        collection_name: str,
        limit: int,
        offset: str = None,
        **payload
    ) -> dict:
    """just retrieve some points which sorted by id

    Args:
        session (ClientSession): the aiohttp client
        collection_name (str): name of the collection
        limit (int): number of the retrieved points
        offset (str, optional): offset for next scroll  Defaults to None.

    Returns:
        dict: result set
    """
    if payload.get('agent', 0) and payload.get('player', 0) and payload.get('ts', 0):
        json = {
            "filter": {
                "must": [
                    {
                        "key": "agent",
                        "match": {
                            "value": payload['agent']
                        }
                    },
                    {
                        "key": "player",
                        "match": {
                            "value": payload['player']
                        }
                    },
                    {
                        "key": "ts",
                        "range": {
                            "gte": payload['ts']
                        }
                    }
                ]
            },
            "limit": limit,
            "offset": offset,
            "with_payload": True
        }
    else:
        json = {
            "limit": limit,
            "offset": offset,
            "with_payload": True
        }
    async with session.post(VECTOR_DB_SCROLL_URL.format(collection_name), json=json) as response:
        data = await response.json(content_type='application/json')
        return data['result']


async def qdrant_count(
        session: ClientSession, 
        collection_name: str, 
        **payload
    ) -> int:
    """count points with custom conditions

    Args:
        session (ClientSession): the aiohttp client
        collection_name (str): name of the collection

    Returns:
        int: points count
    """
    if payload.get('agent', 0) and payload.get('player', 0):
        json = {
            "filter": {
                "must": [
                    {
                        "key": "agent",
                        "match": {
                            "value": payload['agent']
                        }
                    },
                    {
                        "key": "player",
                        "match": {
                            "value": payload['player']
                        }
                    }
                ]
            }
        }
    else:
        json = {}
    async with session.post(VECTOR_DB_COUNT_URL.format(collection_name), json=json) as response:
        data = await response.json(content_type='application/json')
        return data['result']['count']


async def qdrant_list_collections(session: ClientSession) -> List[str]:
    """list all collections

    Args:
        session (ClientSession): the aiohttp client

    Returns:
        List[str]: list of collection names
    """
    async with session.get(VECTOR_DB_GET_COLLECTIONS_URL) as response:
        data = await response.json(content_type='application/json')
        return [x['name'] for x in data['result']['collections']]

```

**Comment by rxy1212** [comment]
With this [doc](https://qdrant.github.io/qdrant/redoc/index.html), theoretically you can do any async request whatever the official [qdrant-client](https://github.com/qdrant/qdrant-client) package provide

**Comment by joein** [comment]
`AsyncQdrantClient` is available as of 1.6.1 with regular interface
